ocd_backend/models/postgres_database.py

```
@@ -0,0 +1,116 @@
+                                    settings.POSTGRES_USERNAME,
+                                    settings.POSTGRES_PASSWORD,
+                                    settings.POSTGRES_HOST,
+                                    settings.POSTGRES_DATABASE)
+        self.engine = create_engine(self.connection_string)
+        self.Session = sessionmaker(bind=self.engine)
+
+    def get_ori_identifier(self, iri):
+        """
+        Retrieves a Resource-based ORI identifier from the database. If no corresponding Resource exists,
+        a new one is created.
+        """
+
+        session = self.Session()
+        try:
+            resource = session.query(Resource).filter(Resource.iri == iri).one()
+            return Uri(Ori, resource.ori_id)
+        except MultipleResultsFound:
+            raise MultipleResultsFound('Multiple resources found for IRI %s' % iri)
+        except NoResultFound:
+            return self.generate_ori_identifier(iri=iri)
+        except:
+            session.rollback()
+            raise
+        finally:
+            session.close()
+
+    def generate_ori_identifier(self, iri):
+        """
+        Generates a Resource with an ORI identifier and adds the IRI as a Source if it does not already exist.
+        """
+
+        session = self.Session()
+        new_id = self.engine.execute(Sequence('ori_id_seq'))
+        new_identifier = Uri(Ori, new_id)
+
+        try:
+            # If the source already exists, create the resource as a child of the source
+            source = session.query(Source).filter(Source.iri == iri).one()
+            source.resources.append(Resource(id=new_id, ori_id=new_id, iri=iri))
+            session.flush()
+        except NoResultFound:
+            # If the source does not exist, create source and resource together
+            source = Source(iri=iri, resources=[Resource(id=new_id, ori_id=new_id, iri=iri)])
+            session.add(source)
+            session.commit()
+        except:
```

PEP8: do not use bare except

---

:+1: